### DISCUSSION TITLE
Are there any ways to register some functions which will be executed when exiting daphne?

### DISCUSSION THREAD

**Discussion by HullQin** [django/daphne#409] [Q&A]
I am using daphne to deploy my websocket app, without django or channels.

I hope the data in memory can be saved on disk when I exit daphne.

Do you have any good ideas? Thanks :)

**Comment by HullQin** [comment]
I thought I got the answer from the source code.

Just write this:

```python3
from twisted.internet import reactor


def before_shutdown():
    print('Succeed!')


reactor.addSystemEventTrigger("before", "shutdown", before_shutdown)


async def application(scope, receive, send):
    pass
```

**Comment by carltongibson** [comment]
You might also use an [ASGI Lifespan middleware](https://asgi.readthedocs.io/en/latest/specs/lifespan.html) for this kind of thing. 🤔

**Reply by HullQin** [reply]
Thanks, it's more graceful ~

**Reply by HullQin** [reply]
Sorry, but it seems that Daphne currently does not support lifespan 😂

**Reply by carltongibson** [reply]
Ah, yes... I thought we'd knocked that one off. ... Still happy to add it with a suitable PR.
